# Pipeline to investigate RNAseq data

A reference-free pipeline to investigate RNAseq data using k-mers and Machine Learning methods to better understand prognosis classification and key mechanisms of AML pathogenesis.

## Required
Kmtricks tool ([git](<https://github.com/tlemane/kmtricks/>))

R version 4.1.2: Package Seqinr ([cran](<https://cran.r-project.org/web/packages/seqinr/index.html>))

C++: Boost String Algorithms Library ([doc](<https://www.boost.org/doc/libs/1_81_0/doc/html/string_algo/reference.html#header.boost.algorithm.string_hpp>))

STAR tool ([git](<https://github.com/alexdobin/STAR>))

Samtools ([git](https://github.com/samtools/samtools>))


## Running
**Kmtricks**

To generate the k-mer count matrix

Necessary files:
- samples : file with samples name and path
- fastq files : files to analyze

```
~/kmtricks pipeline --file samples --run-dir temp --kmer-size 31 -t 4 --until merge
~/kmtricks aggregate --run-dir ./temp --matrix kmer --format text --sorted -t 4 > sorted_matrix.tsv
```

**Adapting matrix script**

To adapt the k-mer count matrix for the next steps

Necessary files:
- samples : file with samples name and path
- sorted_matrix.tsv : matrix generated by Kmtrics

```
./adapt_mat.sh
```

**Coefficient of variation code**

To select the relevant k-mers and generate a fasta file with the k-mers

Necessary files:
- sorted_matrix.tsv.gzip : matrix adapted from last script
- samples_cond : samples name in two lines, each line is one condition

```
./cvar <(zcat sorted_matrix.tsv.gz)
Rscript tsv_to_fasta.R
```

**STAR and Samtools**

To map 

Necessary files:
- output.fasta : fasta file generated from last R script

```
STAR --genomeDir /data/indexes/STAR/2.7.6a/GRCh38.97 --readFilesType Fastx  --readFilesIn output.fa --outFileNamePrefix leucegene_ --outStd Log --runMode alignReads --runThreadN 10 --outSAMunmapped Within
samtools view -b output_Aligned.out.sam |samtools sort > output_Aligned.out.bam	
```

**Getting Annotation**

To get the annotation, and provide a view of k-mers by gene and region

Necessary files:
- output_Aligned.out.bam : file generated from Samtools
- output_Aligned.out.sam : file generated from Samtools

```
Rscript getAnnotation.R
```

**Building train**

To adapt the selected k-mers to train the model

```
Rscript building_train.R
```

**Generating model**

To generate a XGB model with the selected k-mers

Necessary files:
- train.tsv : file from last script

```
./xgb_classification.py
```



Note: examples of the files can be found in the folder *examples*
